lib/utils/workspace-utils.ts

```diff
@@ -0,0 +1,109 @@
+import { fetchApiData } from "helpers/fetchApiData";
+
+export async function createWorkspace({
+  name,
+  description = "",
+  members = [],
+  sessionToken,
+  repos = [],
+}: {
+  name: string;
+  description?: string;
+  members?: any[];
+  sessionToken: string;
+  repos: { full_name: string }[];
+}) {
+  const { data: workspace, error: workspaceError } = await fetchApiData<Workspace>({
+    path: "workspaces",
+    method: "POST",
+    body: { name, description, members },
+    bearerToken: sessionToken,
+    pathValidator: () => true,
+  });
+
+  if (workspace) {
+    const { data: repoData, error: reposError } = await fetchApiData<any[]>({
```

Comment:
@jpmcb, could the `/workspaces` creation endpoint take repos as well? Just asking as I've run into this a few times where the workspace gets created, but then adding repositories fail. Since it's not all in a transaction, the workspace exists, but a retry creates a new workspace when the repos get added successfully in this call, `workspaces/${workspace.id}/repos`.

Reply:
Sure thing: we can add repos directly to workspace creation.

> Since it's not all in a transaction, the workspace exists, but a retry creates a new workspace when the repos get added successfully in this call, `workspaces/${workspace.id}/repos`

I'm confused, why would the retry here create a new workspace if creating a new workspace and adding repos to it are separate?

Reply:
Hmmm I'm second guessing adding a repos payload to workspace creation: what if there are invalid repos in the payload for creating a workspace? Would the whole call fail? Should the workspace still be created but no repos be added? Only the valid repos be added with an error bubbled up for the repos that are invalid?

In my mind, two separate calls (one for creating a workspace and another one for mutating the repos within a workspace) makes logical sense to me.

Reply:
This is the scenario. Say the workspace gets created (first call) and the second call to add repos fails. The workspace has been created, but we're still on the creation page. The user currently sees an error if the workspace (first call) fails or if the repos fail to add (second call).

So say they retry, the create workspace (first call) runs again (successfully), and then the second call finally works, they're redirected to the settings page but now they have an additional workspace (the first one that created the workspace but failed to create the repos in the second call).

I can guard against this by storing the workspaceId  client-side if the first part was successful, but then that seems kind of hacky. This is why I was proposing allowing adding repos on creation. I'd rather it all fail then have one part pass.

The weird thing is what I mentioned to you the other day. Sometimes when the creation of repos fail, they're valid repos (since they come from the search repos endpoint from the GitHub API. Or do we 404 on invalid repos sometimes because a valid repo is a repo that we've indexed with pizza time?

Reply:
It relates to my TODOS above in the PR description.

> - [ ] sometimes creating a repo fails because the endpoint for adding repos to the new workspace gives a 404. The workspace is created successfully, so it does exist. My guess is the endpoint 404s because maybe the database hasn't updated yet, so the repos endpoint 404s. A second attempt could be made as a "temporary" fix
> - [ ] sometimes I get 404s for the updating repos endpoint. Same for getting repos, e.g. 
https://beta.api.opensauced.pizza/v1/workspaces/9e0c76a8-9034-4670-94e8-f73b4a5c8679/repos

Reply:
Sure thing: either way, let's move forward with allowing clients to have a `repos` payload: https://github.com/open-sauced/api/issues/509

I'll stub out some work and get that in today!

---

> Sometimes when the creation of repos fail, they're valid repos

Hmm very weird, still not sure why this would occur: if we can reliably reproduce it, let's add those steps to an issue

Reply:
Let's keep them as is for now as you mentioned it's something we need to sort out in https://github.com/open-sauced/api/issues/511